File: Benchmarks.Tests/RepeaterTests.cs
using System;
using System.Collections.Generic;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Pidgin;
using Stringier;
using Sprache;
using static Benchmarks.Patterns.RepeaterBenchmarks;

namespace Benchmarks {
	[TestClass]
	public class RepeaterTests {
		[TestMethod]
		public void Pidgin() {
			Result<Char, IEnumerable<String>> result;

			result = pidgin.Parse("Hi!Hi!Hi!Hi!Hi!");
			Assert.IsTrue(result.Success);
			Assert.AreEqual("Hi!Hi!Hi!Hi!Hi!", result.Value.Join());

			result = pidgin.Parse("Hi!");
			Assert.IsFalse(result.Success);
		}

		[TestMethod]
		public void Sprache() {
			IResult<IEnumerable<String>> result;

			result = sprache.TryParse("Hi!Hi!Hi!Hi!Hi!");
			Assert.IsTrue(result.WasSuccessful);
			Assert.AreEqual("Hi!Hi!Hi!Hi!Hi!", result.Value.Join());

			result = sprache.TryParse("Hi!");
			Assert.IsFalse(result.WasSuccessful);
		}
	}
}

File: Patterns.Tests/RepeaterTests.fs
namespace Patterns

open System
open Stringier.Patterns
open Microsoft.VisualStudio.TestTools.UnitTesting

[<TestClass>]
type RepeaterTests() =

    [<TestMethod>]
    member _.``consume`` () =
        let pattern = "Hi! " * 2

        ResultAssert.Captures("Hi! Hi! ", pattern.Consume("Hi! Hi! Hi!"))
        ResultAssert.Fails(pattern.Consume("Bye! Hi! "))

    [<TestMethod>]
    member _.``multiplication still works`` () =
        Assert.AreEqual(4, 2 * 2)
        Assert.AreEqual(6.66, 3.33 * 2.0)

    [<TestMethod>]
    member _.``neglect`` () =
        let pattern = not ("Hi!" * 2)
        ResultAssert.Fails(pattern.Consume("Hi!Hi!"))
        ResultAssert.Captures("Oh!Oh!", pattern.Consume("Oh!Oh!"))
        ResultAssert.Captures("Oh!Hi!", pattern.Consume("Oh!Hi!"))